File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/DeepLinkStateMachine.java
/*
 * Copyright (c) 2015-2022 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.snowplow.internal.tracker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.entity.DeepLink;
import com.snowplowanalytics.snowplow.event.DeepLinkReceived;
import com.snowplowanalytics.snowplow.event.Event;
import com.snowplowanalytics.snowplow.internal.constants.TrackerConstants;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
import com.snowplowanalytics.snowplow.tracker.InspectableEvent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class DeepLinkStateMachine implements StateMachineInterface {

    /*
     States: Init, DeepLinkReceived, ReadyForOutput
     Events: DL (DeepLinkReceived), SV (ScreenView)
     Transitions:
      - Init (DL) DeepLinkReceived
      - DeepLinkReceived (SV) ReadyForOutput
      - ReadyForOutput (DL) DeepLinkReceived
      - ReadyForOutput (SV) Init
     Entity Generation:
      - ReadyForOutput
      */

    @NonNull
    @Override
    public List<String> subscribedEventSchemasForTransitions() {
        return Arrays.asList(DeepLinkReceived.SCHEMA, TrackerConstants.SCHEMA_SCREEN_VIEW);
    }

    @NonNull
    @Override
    public List<String> subscribedEventSchemasForEntitiesGeneration() {
        return Arrays.asList(TrackerConstants.SCHEMA_SCREEN_VIEW);
    }

    @NonNull
    @Override
    public List<String> subscribedEventSchemasForPayloadUpdating() {
        return new ArrayList<>();
    }

    @Nullable
    @Override
    public State transition(@NonNull Event event, @Nullable State state) {
        // - Init (DL) DeepLinkReceived
        // - ReadyForOutput (DL) DeepLinkReceived
        if (event instanceof DeepLinkReceived) {
            DeepLinkReceived dlEvent = (DeepLinkReceived) event;
            return new DeepLinkState(dlEvent.url, dlEvent.referrer);
        } else {
            // - Init (SV) Init
            if (state == null) {
                return null;
            }
            // - ReadyForOutput (SV) Init
            DeepLinkState dlState = (DeepLinkState) state;
            if (dlState.readyForOutput) {
                return null;
            }
            // - DeepLinkReceived (SV) ReadyForOutput
            DeepLinkState currentState = new DeepLinkState(dlState.url, dlState.referrer);
            currentState.readyForOutput = true;
            return currentState;
        }
    }

    @Nullable
    @Override
    public List<SelfDescribingJson> entities(@NonNull InspectableEvent event, @Nullable State state) {
        if (state == null) {
            return null;
        }
        DeepLinkState deepLinkState = (DeepLinkState) state;
        if (!deepLinkState.readyForOutput) {
            return null;
        }
        DeepLink entity = new DeepLink(deepLinkState.url)
                .referrer(deepLinkState.referrer);
        return Collections.singletonList(entity);
    }

    @Nullable
    @Override
    public Map<String, Object> payloadValues(@NonNull InspectableEvent event, @Nullable State state) {
        return null;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/statemachine/DeepLinkStateMachine.kt
/*
 * Copyright (c) 2015-2022 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.statemachine

import com.snowplowanalytics.core.constants.TrackerConstants
import com.snowplowanalytics.snowplow.entity.DeepLink
import com.snowplowanalytics.snowplow.event.DeepLinkReceived
import com.snowplowanalytics.snowplow.event.Event
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.tracker.InspectableEvent
import kotlin.collections.ArrayList

class DeepLinkStateMachine : StateMachineInterface {
    /*
     States: Init, DeepLinkReceived, ReadyForOutput
     Events: DL (DeepLinkReceived), SV (ScreenView)
     Transitions:
      - Init (DL) DeepLinkReceived
      - DeepLinkReceived (SV) ReadyForOutput
      - ReadyForOutput (DL) DeepLinkReceived
      - ReadyForOutput (SV) Init
     Entity Generation:
      - ReadyForOutput
      */

    override val identifier: String
        get() = ID

    override val subscribedEventSchemasForTransitions: List<String>
        get() = listOf(DeepLinkReceived.schema, TrackerConstants.SCHEMA_SCREEN_VIEW)

    override val subscribedEventSchemasForEntitiesGeneration: List<String> 
        get() = listOf(TrackerConstants.SCHEMA_SCREEN_VIEW)

    override val subscribedEventSchemasForPayloadUpdating: List<String>
        get() = ArrayList()

    override val subscribedEventSchemasForAfterTrackCallback: List<String>
        get() = emptyList()

    override fun transition(event: Event, state: State?): State? {
        // - Init (DL) DeepLinkReceived
        // - ReadyForOutput (DL) DeepLinkReceived
        return if (event is DeepLinkReceived) {
            DeepLinkState(event.url, event.referrer)
        } else {
            // - Init (SV) Init
            if (state == null) {
                return null
            }
            // - ReadyForOutput (SV) Init
            val dlState = state as? DeepLinkState
            if (dlState?.readyForOutput == true) {
                return null
            }
            // - DeepLinkReceived (SV) ReadyForOutput
            val currentState = dlState?.let { DeepLinkState(it.url, it.referrer) }
            currentState?.readyForOutput = true
            currentState
        }
    }

    override fun entities(event: InspectableEvent, state: State?): List<SelfDescribingJson>? {
        if (state == null) { return null }

        val deepLinkState = state as? DeepLinkState
        if (deepLinkState?.readyForOutput == false) {
            return null
        }
        val entity = deepLinkState?.let { DeepLink(it.url).referrer(it.referrer) }
        return entity?.let { listOf<SelfDescribingJson>(entity) }
    }

    override fun payloadValues(event: InspectableEvent, state: State?): Map<String, Any>? {
        return null
    }

    override fun afterTrack(event: InspectableEvent) {
    }

    companion object {
        val ID: String
            get() = "DeepLinkContext"
    }
}
